presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotFilterExpressionConverter.java

```diff
@@ -89,15 +100,103 @@ private PinotExpression handleLogicalBinary(
         }
         List<RowExpression> arguments = call.getArguments();
         if (arguments.size() == 2) {
+            // Trying to pushdown the predicate of time field comparing to a time literal.
+            if (isDateOrTimestampExpression(arguments)) {
+                return handleDateOrTimestampBinaryExpression(operator, arguments, context);
+            }
             return derived(format(
-                    "(%s %s %s)",
-                    arguments.get(0).accept(this, context).getDefinition(),
-                    operator,
-                    arguments.get(1).accept(this, context).getDefinition()));
+                "(%s %s %s)",
+                arguments.get(0).accept(this, context).getDefinition(),
+                operator,
+                arguments.get(1).accept(this, context).getDefinition()));
         }
         throw new PinotException(PINOT_UNSUPPORTED_EXPRESSION, Optional.empty(), format("Unknown logical binary: '%s'", call));
     }
 
+    private PinotExpression handleDateOrTimestampBinaryExpression(String operator, List<RowExpression> arguments, Function<VariableReferenceExpression, Selection> context)
+    {
+        String left = arguments.get(0).accept(this, context).getDefinition();
+        String right = arguments.get(1).accept(this, context).getDefinition();
+        left = handleTimeStampAndDateCast(arguments.get(1), left);
+        right = handleTimeStampAndDateCast(arguments.get(0), right);
+        return derived(format("(%s %s %s)", left, operator, right));
+    }
+
+    private boolean isDateOrTimestampExpression(List<RowExpression> arguments)
+    {
+        for (RowExpression rowExpression : arguments) {
```

Comment:
How about naming rowExpression to 'argument' ?

Can you please add a Preconditions.checkState that arguments.size() == 2, I think this is only called in the binary codepath  ? 

I didn't follow why we return success right away when the first argument matches. It seems you want exactly one of these two cases: 
- Left should be a cast of the valid type, right should be a literal (ConstantExpression)
- Right should be a cast of the valid type, Left should be a literal (ConstantExpression)

Whereas this function is being a bit more permissive than that. The reason I am concerned about that permissiveness is because then we will eventually crap out at line 185 or 194 below. Is it possible to tighten the check ?

Reply:
Yes, refined this code block and add comments.